File: server/ruby/app/infrastructure/mappers/payment_details/three_d_secure_results_mapper.rb
module Infrastructure
  module Mappers
    module PaymentDetails
      class ThreeDSecureResultsMapper
        def self.from_sdk_response(response)
          return nil if response.nil?

          Business::Domain::Payments::PaymentDetails::ThreeDSecureResults.new(
            acs_transaction_id: response.acs_transaction_id,
            applied_exemption: response.applied_exemption,
            authentication_status: response.authentication_status,
            cavv: response.cavv,
            challenge_indicator: response.challenge_indicator,
            ds_transaction_id: response.ds_transaction_id,
            eci: response.eci,
            exemption_engine_flow: response.exemption_engine_flow,
            flow: response.flow,
            liability: response.liability,
            scheme_eci: response.scheme_eci,
            version: response.version,
            xid: response.xid
          )
        end
      end
    end
  end
end

File: server/python/app/infrastructure/mappers/payment_details/three_d_secure_results_mapper.py
from typing import Optional
from app.application.domain.payments.payment_details.three_d_secure_results import ThreeDSecureResults
from onlinepayments.sdk.domain.three_d_secure_results import ThreeDSecureResults as ThreeDSecureResultsSdk

class ThreeDSecureResultsMapper:

    @staticmethod
    def map_from_sdk_response(response: Optional[ThreeDSecureResultsSdk]) -> Optional[ThreeDSecureResults]:
        if response is None:
            return None

        dto = ThreeDSecureResults()
        dto.acs_transaction_id = response.acs_transaction_id
        dto.applied_exemption = response.applied_exemption
        dto.authentication_status = response.authentication_status
        dto.cavv = response.cavv
        dto.challenge_indicator = response.challenge_indicator
        dto.ds_transaction_id = response.ds_transaction_id
        dto.eci = response.eci
        dto.exemption_engine_flow = response.exemption_engine_flow
        dto.flow = response.flow
        dto.liability = response.liability
        dto.scheme_eci = response.scheme_eci
        dto.version = response.version
        dto.xid = response.xid

        return dto